File: backend/src/main/java/net/archasmiel/skufapi/controller/AuthController.java
package net.archasmiel.skufapi.controller;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.archasmiel.skufapi.api.exception.ErrorResponse;
import net.archasmiel.skufapi.api.response.ApiResponse;
import net.archasmiel.skufapi.api.request.auth.GoogleAuthRequest;
import net.archasmiel.skufapi.api.response.auth.MeResponse;
import net.archasmiel.skufapi.api.request.auth.LoginRequest;
import net.archasmiel.skufapi.api.request.auth.RegisterRequest;
import net.archasmiel.skufapi.security.SecurityContext;
import net.archasmiel.skufapi.service.AuthenticationService;
import net.archasmiel.skufapi.service.RegistrationService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.bind.annotation.*;

@Slf4j
@RestController
@RequiredArgsConstructor
@RequestMapping("/api/auth")
public class AuthController {

    private final AuthenticationService authService;
    private final RegistrationService regService;

    @PostMapping("/register")
    public ResponseEntity<ApiResponse> signup(@RequestBody RegisterRequest request) {
        try {
            ApiResponse response = regService.signUp(request);
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            ErrorResponse response = new ErrorResponse("SIGNUP_FAILED", e.getMessage());
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(response);
        }
    }

    @PostMapping("/login")
    public ResponseEntity<ApiResponse> login(@RequestBody LoginRequest request) {
        try {
            ApiResponse response = authService.signIn(request);
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            ErrorResponse response = new ErrorResponse("AUTH_FAILED", e.getMessage());
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(response);
        }
    }

    @PostMapping("/google")
    public ResponseEntity<ApiResponse> loginGoogle(@RequestBody GoogleAuthRequest request) {
        try {
            ApiResponse response = authService.signIn(request);
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            ErrorResponse response = new ErrorResponse("AUTH_FAILED", e.getMessage());
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(response);
        }
    }

    @GetMapping("/me")
    public ResponseEntity<ApiResponse> getMe(@SecurityContext Authentication auth) {
        try {
            if (auth == null) {
                throw new RuntimeException("No authentication");
            }
            if (auth.getPrincipal() != null && auth.getPrincipal() instanceof UserDetails details) {
                ApiResponse response =
                        new MeResponse(details.getUsername(), details.getAuthorities().toString());
                return ResponseEntity.ok(response);
            }
            throw new RuntimeException("Null or empty user principal");
        } catch (Exception e) {
            ErrorResponse response = new ErrorResponse("REQUEST_FAILED", e.getMessage());
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(response);
        }
    }

    @Deprecated
    @GetMapping("/logout")
    public ResponseEntity<ApiResponse> logout(@SecurityContext Authentication auth) {
        try {
            if (auth == null) {
                throw new RuntimeException("No authentication");
            }
            if (auth.getPrincipal() != null && auth.getPrincipal() instanceof UserDetails details) {
                ApiResponse response =
                        new MeResponse(details.getUsername(), "Successfully logged out");
                return ResponseEntity.ok(response);
            }
            throw new RuntimeException("Null or empty user principal");
        } catch (Exception e) {
            ErrorResponse response = new ErrorResponse("REQUEST_FAILED", e.getMessage());
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(response);
        }
    }
}

File: backend/src/main/kotlin/net/archasmiel/skufapi/controller/AuthController.kt
package net.archasmiel.skufapi.controller

import net.archasmiel.skufapi.api.request.auth.GoogleAuthRequest
import net.archasmiel.skufapi.api.request.auth.LoginRequest
import net.archasmiel.skufapi.api.request.auth.RegisterRequest
import net.archasmiel.skufapi.api.response.ApiResponse
import net.archasmiel.skufapi.api.response.auth.MeResponse
import net.archasmiel.skufapi.exception.auth.AuthenticationException
import net.archasmiel.skufapi.exception.token.GoogleTokenException
import net.archasmiel.skufapi.exception.token.JwtTokenException
import net.archasmiel.skufapi.exception.user.GoogleUserExistException
import net.archasmiel.skufapi.exception.user.UserExistException
import net.archasmiel.skufapi.security.SecurityContext
import net.archasmiel.skufapi.service.AuthenticationService
import net.archasmiel.skufapi.service.RegistrationService
import org.springframework.http.ResponseEntity
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/auth")
class AuthController(
    private val authService: AuthenticationService,
    private val registerService: RegistrationService
) {

    @Throws(
        JwtTokenException::class,
        UserExistException::class)
    @PostMapping("/register")
    fun signup(@RequestBody request: RegisterRequest): ResponseEntity<ApiResponse> {
        return ResponseEntity.ok(
            registerService.signUp(request)
        )
    }

    @Throws(
        AuthenticationException::class,
        UsernameNotFoundException::class,
        JwtTokenException::class,
        GoogleUserExistException::class)
    @PostMapping("/login")
    fun login(@RequestBody request: LoginRequest): ResponseEntity<ApiResponse> {
        return ResponseEntity.ok(
            authService.signIn(request)
        )
    }

    @Throws(
        UserExistException::class,
        JwtTokenException::class,
        GoogleTokenException::class)
    @PostMapping("/google")
    fun loginGoogle(@RequestBody request: GoogleAuthRequest): ResponseEntity<ApiResponse> {
        return ResponseEntity.ok(
            authService.signIn(request)
        )
    }

    @GetMapping("/me")
    fun getMe(@SecurityContext user: UserDetails): ResponseEntity<ApiResponse> {
        return ResponseEntity.ok(
            MeResponse(user.username, "Authorities: ${user.authorities}")
        )
    }

    @GetMapping("/logout")
    fun logout(@SecurityContext user: UserDetails): ResponseEntity<ApiResponse> {
        return ResponseEntity.ok(
            MeResponse(user.username, "Successfully logged out")
        )
    }
}